net/portmapper/upnp.go

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+package portmapper
+
+import (
+	"context"
+	"time"
+
+	"github.com/tailscale/goupnp/dcps/internetgateway2"
+	"golang.org/x/sync/errgroup"
+	"inet.af/netaddr"
+)
+
+// References:
+//
+// WANIP Connection v2: http://upnp.org/specs/gw/UPnP-gw-WANIPConnection-v2-Service.pdf
+
+type upnpMapping struct {
+	gw       netaddr.IP
+	external netaddr.IPPort
+	internal netaddr.IPPort
+	useUntil time.Time
+	client   upnpClient
+}
+
+func (u *upnpMapping) isCurrent() bool                { return u.useUntil.After(time.Now()) }
+func (u *upnpMapping) validUntil() time.Time          { return u.useUntil }
+func (u *upnpMapping) externalIPPort() netaddr.IPPort { return u.external }
+func (u *upnpMapping) release() {
+	u.client.DeletePortMapping(context.Background(), "", u.external.Port(), "udp")
+}
+
+// UPnP Client is an interface over the multiple different clients exported by goupnp,
+// exposing the functions we need for portmapping. They are auto-generated from XML-specs.
+type upnpClient interface {
+	AddPortMapping(
+		ctx context.Context,
+		remoteHost string,
+		externalPort uint16,
+		protocol string,
+		internalPort uint16,
+		internalClient string,
+		enabled bool,
+		portMappingDescription string,
+		leaseDuration uint32,
```

Comment:
duration in what? Go code should use time.Duration in general. But if we must use something else, it needs to say the unit.